python/dgl/heterograph.py

```diff
@@ -0,0 +1,1214 @@
+          ``dst_type_name``, and edge type name ``edge_type_name``.
+
+        Parameters
+        ----------
+        key : str or tuple
+            See above
+
+        Returns
+        -------
+        DGLBaseHeteroGraph
+            The induced subgraph.
+        """
+        pass
+
+
+class DGLHeteroGraph(DGLBaseHeteroGraph):
+    """Base heterogeneous graph class.
+
+    The graph stores nodes, edges and also their (type-specific) features.
+
+    Heterogeneous graphs are by default multigraphs.
+
+    Parameters
+    ----------
+    metagraph, node_types, edge_connections_by_type :
+        See DGLBaseHeteroGraph
+    node_frame : dict[str, FrameRef], optional
+        Node feature storage per type
+    edge_frame : dict[str, FrameRef], optional
+        Edge feature storage per type
+    readonly : bool, optional
+        Whether the graph structure is read-only (default: False)
+    """
+    # pylint: disable=unused-argument
+    def __init__(
+            self,
+            metagraph,
+            node_types,
+            edge_connections_by_type,
+            node_frame=None,
+            edge_frame=None,
+            readonly=False):
+        super(DGLHeteroGraph, self).__init__(metagraph, node_types, edge_connections_by_type)
+
+    # TODO: REVIEW
+    def add_nodes(self, num, node_type, data=None):
+        """Add multiple new nodes of the same node type
+
+        Parameters
+        ----------
+        num : int
+            Number of nodes to be added.
+        node_type : str
+            Type of the added nodes.  Must appear in the metagraph.
+        data : dict, optional
+            Feature data of the added nodes.
+
+        Examples
+        --------
+        The variable ``g`` is constructed from the example in
+        DGLBaseHeteroGraph.
+
+        >>> g['game'].number_of_nodes()
+        2
+        >>> g.add_nodes(3, 'game')  # add 3 new games
+        >>> g['game'].number_of_nodes()
+        5
+        """
+        pass
+
+    # TODO: REVIEW
+    def add_edge(self, u, v, utype, vtype, etype, data=None):
+        """Add an edge of ``etype`` between u of type ``utype`` and v of type
+        ``vtype``.
+
+        Parameters
+        ----------
+        u : int
+            The source node ID of type ``utype``.  Must exist in the graph.
+        v : int
+            The destination node ID of type ``vtype``.  Must exist in the
+            graph.
+        utype : str
+            The source node type name.  Must exist in the metagraph.
+        vtype : str
+            The destination node type name.  Must exist in the metagraph.
+        etype : str
+            The edge type name.  Must exist in the metagraph.
+        data : dict, optional
+            Feature data of the added edge.
+
+        Examples
+        --------
+        The variable ``g`` is constructed from the example in
+        DGLBaseHeteroGraph.
+
+        >>> g['user', 'game', 'plays'].number_of_edges()
```

Comment:
Is it possible to directly call `g['plays']`(assume no semantic ambiguity)? `g['user', 'game', 'plays']` seems too long?

Reply:
multiple edge types can have the same edge type name

Reply:
More accurately, edge types could share the same names *while* connecting to different node types.  For instance:
```
Piano <---(Plays)--- User ---(Plays)---> Game
```
The problem is: do we really want to allow different edge types having the same name?  Or should we force the users to have distinct names for each edge type (like, **force** the users to rename `Plays` to `Plays-piano` and `Plays-game`)?

Reply:
Another problem is `g['plays']` returns edge subgraph while `g['user']` returns node induced subgraph. This makes the type system inconsistent.

Reply:
> More accurately, edge types could share the same names _while_ connecting to different node types. For instance:
> 
> ```
> Piano <---(Plays)--- User ---(Plays)---> Game
> ```
> 
> The problem is: do we really want to allow different edge types having the same name? Or should we force the users to have distinct names for each edge type (like, **force** the users to rename `Plays` to `Plays-piano` and `Plays-game`)?

In this case, forcing unique name can make the code more concise and I'm for it. My concern is something like:
```
Piano <---(have)--- User ---(have)---> Game
```
Will user have this kind of relation and want to do the computation on all `have` edges?

Reply:
> My concern is something like:
> 
> ```
> Piano <---(have)--- User ---(have)---> Game
> ```
> 
> Will user have this kind of relation and want to do the computation on all have edges?

We assume not; they need to compute one edge type at a time (the idea itself is debatable; see discussion in L659)

EDIT: I accidentally replied by editing the previous comment...